src/MyNUnitImplementation/MyNUnitImplementation/MyNUnit.cs

```diff
@@ -0,0 +1,286 @@
+            }
+
+            /// <summary>
+            /// Creates a report from class log
+            /// </summary>
+            /// <returns> Report</returns>
+            public ClassTestingReport MakeReport() => new ClassTestingReport(testClass, classLog);
+
+            /// <summary>
+            /// Run all marked methods
+            /// </summary>
+            public void Run()
+            {
+                if (RunAfterOrBeforeClassMethod(methods.BeforeClass, true))
+                {
+                    return;
+                }
+
+                Parallel.ForEach(methods.Tests, RunTestMethod);
+                RunAfterOrBeforeClassMethod(methods.AfterClass, false);
+            }
+
+            /// <summary>
+            /// Run after or before class method
+            /// </summary>
+            /// <param name="method">Test method</param>
+            /// <param name="isBefore"> Flag for before method</param>
+            private bool RunAfterOrBeforeClassMethod(MethodInfo? method, bool isBefore)
+            {
+                try
+                {
+                    method?.Invoke(null, null);
+                }
+                catch (Exception e)
+                {
+                    if (isBefore)
+                    {
+                        Parallel.ForEach(methods.Tests,
+                            test =>
+                            {
+                                classLog.Add(new TestInfo(test.Name, TestStatus.Errored,
+                                    reason: $"Errored because of after class method falling: {e.Message}"));
+                            });
+                    }
+                    else
+                    {
+                        Parallel.ForEach(classLog, log =>
+                        {
+                            log.Status = TestStatus.Errored;
+                            log.Reason = $"Errored because of after class method falling: {e.Message}";
+                        });
+                    }
+
+                    return true;
+                }
+
+                return false;
+            }
+
+            private bool RunAfterOrBeforeMethod(ConcurrentBag<MethodInfo> after, string name, object instance,
+                bool IsBefore)
+            {
+                try
+                {
+                    foreach (var method in after)
+                    {
+                        method.Invoke(instance, null);
+                    }
+                }
+                catch (Exception e)
+                {
+                    if (IsBefore)
+                    {
+                        classLog.Add(new TestInfo(name, TestStatus.Errored,
+                            reason: $"Errored because of before method falling: {e.Message}"));
+                    }
+                    else
+                    {
+                        Parallel.ForEach(classLog, log =>
+                        {
+                            if (log.MethodName != name) return;
```

Comment:
```suggestion
                            if (log.MethodName != name) 
                            {
                                return;
                            }
```
Это делает поток управления более явным